### PR TITLE
8258953: AArch64: move NEON instructions to aarch64_neon.ad

### PR DESCRIPTION
**Pull request by dgbo**
As discussed in [1], all NEON instructions should be moved from `aarch64.ad` to `aarch64_neon.ad`.

In the first commit [2] of this PR, the NEON instructions are deleted from `aarch64.ad` and appended to `aarch64_neon.ad`.
I compared the generated code in `aarch64_neon.ad` with original code in `aarch64.ad`, no suspicious differences found.
The last two commits just simply move code around in `aarch64_neon.ad` to put related instructions together, i.e. `LoadStore` [3], `Reduction` [4].

This also supports vector length 4 for `vsraa8B_imm` and `vsrla8B_imm`, vector length 2 for `vsraa4S_imm` and `vsrla4S_imm`, fixes few typos, e.g. `vor8B`, `vsrla4S_imm`.

[1] https://github.com/openjdk/jdk/pull/1215#issuecomment-728186803
[2] https://github.com/dgbo/jdk/commit/40cbe99e647cdf93712edf8f77ab3b5b30ea0a95
[3] https://github.com/dgbo/jdk/commit/695fb8f8ef009b733a8f804e791347f4bfe2572e
[4] https://github.com/dgbo/jdk/commit/e0c38aa9aaa6af9925a3821328384b1e2b2c2070

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258953](https://bugs.openjdk.java.net/browse/JDK-8258953): AArch64: move NEON instructions to aarch64_neon.ad

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2273/head:pull/2273`
`$ git checkout pull/2273`

### CODE REVIEW

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 1]
```diff
@@ -3702,3 +3996,2150 @@ instruct vabd2D(vecX dst, vecX src1, vecX src2)
...
+
+// --------------------------------- OR ---------------------------------------
+
+instruct vor8B(vecD dst, vecD src1, vecD src2)
+%{
+  predicate(n->as_Vector()->length_in_bytes() == 4 ||
+            n->as_Vector()->length_in_bytes() == 8);
+  match(Set dst (OrV src1 src2));
+  ins_cost(INSN_COST);
+  format %{ "orr  $dst,$src1,$src2\t# vector (8B)" %}
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5022]: Good catch for this typo!

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 2]
```diff
@@ -3702,3 +3996,2150 @@ instruct vabd2D(vecX dst, vecX src1, vecX src2)
...
+  ins_cost(INSN_COST);
+  format %{ "usra    $dst, $src, $shift\t# vector (4H)" %}
+  ins_encode %{
+    int sh = (int)$shift$$constant;
+    if (sh >= 16) {
+      __ eor(as_FloatRegister($src$$reg), __ T8B,
+             as_FloatRegister($src$$reg),
+             as_FloatRegister($src$$reg));
+    } else {
+      __ usra(as_FloatRegister($dst$$reg), __ T4H,
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5937]: I see you have fixed this typo, from ushr to usra. I presume original version generates wrong code and produces wrong results for specific case? If so, do you think it deserves a separate fix, e.g. for jdk16?

**Review by nsjian — COMMENTED** [review]
I managed to sort all the instructs and compare them with and without the patch. They are general the same except for some trailing whitespaces and typos you mentioned.

**Comment by dgbo** [comment]
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> 
> On 1/28/21 10:40 AM, Ningsheng Jian wrote:
> 
> > I see you have fixed this typo, from ushr to usra. I presume original version generates wrong code and produces wrong results for specific case? If so, do you think it deserves a separate fix, e.g. for jdk16?
> 
> It does. This patch should change nothing at all, except moving
> text from A to B.
> 
> --
> Andrew Haley (he/him)
> Java Platform Lead Engineer
> Red Hat UK Ltd. <https://www.redhat.com>
> https://keybase.io/andrewhaley
> EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

@nsjian @theRealAph Thank you for the comments. I'll raise a seperate PR to fix this right now.

BTW, since Andrew says we should change nothing at all in this move, do you think we should also do the things below in separtate PRs?
1. fix the typo of `vor8B`.
2. supporting vector length 4 for `vsraa8B_imm` and `vsrla8B_imm`, vector length 2 for `vsraa4S_imm` and `vsrla4S_imm`.

**Comment by dgbo** [comment]
Updated. The whitespaces mentioned are addressed.
The format typo fix in `vor8B` is kept, other instructions are appended to aarch64_neon.ad.

**Review by theRealAph — APPROVED** [review]
That looks fine. I haven't been able to check that all this patch does is move code from aarch64.ad to aarch64_neon.ad, but I believe you.

**Review by nsjian — APPROVED** [review]
I compared all-ad-src.ad with and without the patch, and it looked good to me.

**Comment by dgbo** [comment]
Thanks for the review.
/integrate

**Pull request closed without merging**